src/solve/fulfill.rs

```diff
@@ -180,87 +202,82 @@ impl<'s> Fulfill<'s> {
+            // and can return the results of inference as a unique substitution
+            let constraints = self.constraints.into_iter().collect();
+            let constrained = self.infer.canonicalize(&ConstrainedSubst { subst, constraints });
+            Ok(Solution::Unique(constrained.quantified))
         } else {
-            debug!("still have {} ambiguous obligations: {:#?}",
-                   self.obligations.len(), self.obligations);
-            Ok(Successful::Maybe)
-        }
-    }
-
-    fn solve_one(&mut self,
-                 wc: &InEnvironment<WhereClauseGoal>,
-                 inference_progress: &mut bool)
-                 -> Result<Successful> {
-        debug!("fulfill::solve_one(wc={:?})", wc);
-
-        let quantified_wc = self.infer.make_query(&wc);
-        let solution = self.solver.solve(quantified_wc.clone())?;
-
-        // Regardless of whether the result is ambiguous or not,
-        // solving the where-clause may have yielded a refined
-        // goal. For example, if the original where-clause was
-        // something like `Foo<?4>: Borrow<?3>`, we would have
-        // "quantified" that to yield `exists ?0, ?1. Foo<?0>: Borrow<?1>`.
-        // We may now have gotten back a refined goal like `exists ?0. Foo<?0>:
-        // Borrow<Foo<?0>>`. In that case, we can unify `?3` with `Foo<?4>`.
-        //
-        // To make that unification happen, we first instantiate all
-        // the variables on the goal we got back with new inference
-        // variables. So we would thus convert `exists ?0. Foo<?0>:
-        // Borrow<Foo<?0>>` into `Foo<?5>: Borrow<Foo<?5>>`.  We would
-        // then unify this with our original goal (`Foo<?4>:
-        // Borrow<?3>`). This will result in the equations `?4 = ?5`
-        // and `?3 = Foo<?5>`.
-        //
-        // As a potential efficiency improvement, one could imagine a
-        // more algorithm written just for this case instead of
-        // instantiating with variables and applying the standard
-        // unification algorithm. But this is good enough for now.
-        let new_type_info = {
-            solution.refined_goal.binders != quantified_wc.binders ||
-            solution.refined_goal.value.value != quantified_wc.value
-        };
+            // If we have obligations remaining, we need to determine how to
+            // package up what we learned about type inference as an ambiguous
+            // solution.
 
-        debug!("fulfill::solve_one: new_type_info={}", new_type_info);
+            if subst.is_trivial(&mut self.infer) {
+                // In this case, we didn't learn *anything* definitively. So
+                // now, we go one last time through the obligations, this time
+                // applying even *tentative* inference suggestions, so that we
+                // can yield these upwards as our own suggestions. In
+                // particular, we yield up the first one we can find.
```

Comment:
First in what sense? we're popping the obligations, after all, so this isn't the first we would have encountered had we been applying the substs "in order", right? (Or am I missing something?)